### DISCUSSION TITLE
[巨坑] 隐藏的ViewModel？

### DISCUSSION THREAD

**Discussion by AlvinRey** [BYJRK/DotNet-Discussions#200] [WPF 开发]
这个问题我是在Avalonia的项目上首次发现的，大家可以试试在WPF上能不能复现。
OK废话不多说，直接上问题。
在MainWindow的后置代码的构造函数中，定义一个MainWindowViewModel，并将其赋值给DataContext。
![83a58625c0475d89361a01b20f95f7d1](https://github.com/BYJRK/DotNet-Discussions/assets/56423507/a80dca93-08a9-4fb0-9ff0-cd91cc15e624)
然后在MainWindow中有两个按钮，一个按钮的Click绑定后置代码中的OnClick函数，另一个按钮的Command绑定MainWindowViewModel中的ClickCommand。
![764b5bd5cbc34ee5fd75e3d2f427d1bb](https://github.com/BYJRK/DotNet-Discussions/assets/56423507/23e18556-2f8a-48cf-861a-286f6bd72064)
![3c21aed3b450d552d758713e036f85e5](https://github.com/BYJRK/DotNet-Discussions/assets/56423507/fd33d018-8e77-45e2-b863-5cf040b15243)
![19a28af19de6c35bc8de5b10a4a1cc64](https://github.com/BYJRK/DotNet-Discussions/assets/56423507/d6376e17-2a86-403f-9d83-e0e3486e381b)
当点击这个两个按钮时，输出的HashCode不一致，证明是两个不同的ViewModel。
![7290160f7a70d9932ec84cfe0a84ec71](https://github.com/BYJRK/DotNet-Discussions/assets/56423507/b0a502e3-99b0-448a-a67d-4efea10d2d0f)
为什么会这样？

当初在 #198 上遇到过相关的问题，开始以为是线程不同导致的，但这次我也输出了线程，证明是不是线程的问题，而是根本就是两个不同的ViewModel！！所以才导致 #198 中更新ObsverableCollection后UI不更新的问题。

**Comment by AlvinRey** [comment]
OK,问题解决！是Avalonia的MVVM模板的原因。在模板默认的App.axmal.cs中，在指定程序的MainWindow时，会顺带给其DataContext初始化一个ViewModel，加上我自己在后置代码中又new了一个，这样就会导致出现两个不同的ViewModel。
![image](https://github.com/BYJRK/DotNet-Discussions/assets/56423507/9cb68090-5b91-4fc4-a494-63b5ca0d7cd9)

**Reply by BYJRK** [reply]
你是不是用了 Avalonia 的 MVVM 模板？App 里面的代码应该是用了这个模板才会生成的吧？我没怎么用过，我一般是创建空项目
